modules/beefy/src/lib.rs

```diff
@@ -0,0 +1,652 @@
+		#[pallet::constant]
+		type CommitmentsToKeep: Get<u32>;
+
+		/// The chain we are bridging to here.
+		type BridgedChain: ChainWithBeefy;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T, I = ()>(PhantomData<(T, I)>);
+
+	#[pallet::hooks]
+	impl<T: Config<I>, I: 'static> Hooks<BlockNumberFor<T>> for Pallet<T, I> {
+		fn on_initialize(_n: T::BlockNumber) -> frame_support::weights::Weight {
+			<RequestCount<T, I>>::mutate(|count| *count = count.saturating_sub(1));
+
+			(0_u64)
+				.saturating_add(T::DbWeight::get().reads(1))
+				.saturating_add(T::DbWeight::get().writes(1))
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> OwnedBridgeModule<T> for Pallet<T, I> {
+		const LOG_TARGET: &'static str = LOG_TARGET;
+		type OwnerStorage = PalletOwner<T, I>;
+		type OperatingMode = BasicOperatingMode;
+		type OperatingModeStorage = PalletOperatingMode<T, I>;
+	}
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I>
+	where
+		BridgedMmrHashing<T, I>: 'static + Send + Sync,
+	{
+		/// Initialize pallet with BEEFY authority set and best known finalized block number.
+		#[pallet::weight((T::DbWeight::get().reads_writes(2, 3), DispatchClass::Operational))]
+		pub fn initialize(
+			origin: OriginFor<T>,
+			init_data: InitializationDataOf<T, I>,
+		) -> DispatchResult {
+			Self::ensure_owner_or_root(origin)?;
+
+			let is_initialized = <ImportedCommitmentsInfo<T, I>>::exists();
+			ensure!(!is_initialized, <Error<T, I>>::AlreadyInitialized);
+
+			log::info!(target: LOG_TARGET, "Initializing bridge BEEFY pallet: {:?}", init_data);
+			Ok(initialize::<T, I>(init_data)?)
+		}
+
+		/// Change `PalletOwner`.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_owner(origin: OriginFor<T>, new_owner: Option<T::AccountId>) -> DispatchResult {
+			<Self as OwnedBridgeModule<_>>::set_owner(origin, new_owner)
+		}
+
+		/// Halt or resume all pallet operations.
+		///
+		/// May only be called either by root, or by `PalletOwner`.
+		#[pallet::weight((T::DbWeight::get().reads_writes(1, 1), DispatchClass::Operational))]
+		pub fn set_operating_mode(
+			origin: OriginFor<T>,
+			operating_mode: BasicOperatingMode,
+		) -> DispatchResult {
+			<Self as OwnedBridgeModule<_>>::set_operating_mode(origin, operating_mode)
+		}
+
+		/// Submit a commitment generated by BEEFY authority set.
+		///
+		/// It will use the underlying storage pallet to fetch information about the current
+		/// authority set and best finalized block number in order to verify that the commitment
+		/// is valid.
+		///
+		/// If successful in verification, it will update the underlying storage with the data
+		/// provided in the newly submitted commitment.
+		#[pallet::weight(0)]
+		pub fn submit_commitment(
+			origin: OriginFor<T>,
+			commitment: BridgedBeefySignedCommitment<T, I>,
+			validator_set: BridgedBeefyValidatorSet<T, I>,
```

Comment:
IIUC, according to the `verify_commitment`, this should be the current authorities set, right? Why do we need it if we're already storing it under the `CurrentAuthoritySet` key?

Reply:
(I don't see any places where `CurrentAuthoritySet` is updated actually - do you plan to fix it in follow up PRs?)

Reply:
Sorry, I must admit that the names `authority_set` and `validator_set` are confusing. I need to find better ones. Maybe `AuthoritySetInfo`/`AuthoritySetMetadata` and `AuthoritySet`/`FullAuthoritySet` would be better.

The authority set contains only some metadata about the validators + the MMR root. It's defined [here](https://github.com/paritytech/substrate/blob/49734dd1d72a00b9d3b87ba397661a63e0e17af3/primitives/beefy/src/mmr.rs#L105-L124).

The validator set is the full list of authority ids (public keys). It's defined [here](https://github.com/paritytech/substrate/blob/49734dd1d72a00b9d3b87ba397661a63e0e17af3/primitives/beefy/src/lib.rs#L124-L129).

We're only storing the `AuthoritySet` (metadata + MMR root) and we receive the `ValidatorSet` (full list of authority public keys). Then:
1. we verify the provided validators are correct by checking `ValidatorSet` against the stored `AuthoritySet::root`
2. we use the provided validators public keys to verify the commitment signatures

The `CurrentAuthoritySet` is updated [here](https://github.com/serban300/parity-bridges-common/blob/64cae305e2a79d5b070927995cf6646e93e3f032/modules/beefy/src/lib.rs#L240-L242)

Reply:
Ok, thanks for the explanation :) Yeah - maybe renaming to something else is a good idea :)

Reply:
Done.

Reply:
I also suggest to think about optimizing usage of this argument a bit. If you have chose not to submit proof-of-membership and instead submit the whole set, then there are many low-hanging optimizations that will save a lot of tx/block size (thinking of bridge hub - would allow other transactions to be included into block). E.g. we may only submit whole set and verify merkle proof when authorities set is changed (that'ss what has been implemented in #1367 initially) or something like that (e.g. cache some public keys when they're first verified and then relayer may only submit public keys that are missing from the runtime storage). This is an optimization for a follow-up PRs of course but worth adding TODO or filing an issue for that.

Reply:
Makes sense. Thanks for the suggestions ! I created #1608 to track this.